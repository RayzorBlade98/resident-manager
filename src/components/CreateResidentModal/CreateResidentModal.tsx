import React, { useState } from "react";
import { Resident, emptyResident } from "_/types/resident";
import { addResident } from "_/states/saveStates/resident_state";
import GenericModal from "../GenericComponents/GenericModal/GenericModal";
import { Button, TextField } from "@mui/material";

export interface CreateResidentModalProps {
  show: boolean;
  onClose: () => void;
}

function CreateResidentModal(props: CreateResidentModalProps): JSX.Element {
  const [resident, setResident] = useState<Resident>(emptyResident());

  function residentUpdater(field: string) {
    function updateResident(event: React.ChangeEvent<HTMLInputElement>): void {
      setResident({
        ...resident,
        [field]: event.target.value,
      });
    }
    return updateResident;
  }

  function onSave(): void {
    addResident(resident);
    props.onClose();
  }

  return (
    <GenericModal title="Neuer Mieter" {...props}>
      {/*Body*/}
      <form>
        <div className="row mb-4">
          <div className="col">
            <TextField
              id="firstName"
              label="Vorname"
              variant="outlined"
              required
              onChange={residentUpdater("firstName")}
            />
          </div>
          <div className="col">
            <TextField
              id="lastName"
              label="Nachname"
              variant="outlined"
              required
              onChange={residentUpdater("lastName")}
            />
          </div>
        </div>
      </form>
      {/*Footer*/}
      <>
        <Button variant="contained" onClick={onSave}>
          Erstellen
        </Button>
      </>
    </GenericModal>
  );
}

export default CreateResidentModal;
