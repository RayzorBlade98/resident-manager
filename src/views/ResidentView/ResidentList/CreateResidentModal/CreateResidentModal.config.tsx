import HomeIcon from '@mui/icons-material/Home';
import KeyIcon from '@mui/icons-material/Key';
import PersonIcon from '@mui/icons-material/Person';
import React from 'react';
import { ValidationConstraint } from '../../../../utils/validation/constraints';
import Validator from '../../../../utils/validation/validator';
import MonthYear from '_/extensions/date/month_year.extension';
import Apartment from '_/models/property/apartment';
import { ContractResident } from '_/models/resident/contractResident';
import { FormConfig } from '_/types/FormConfig';
import { CurrencyInCents } from '_/utils/currency/currency.utils';

/**
 * All values that can be submitted in the form
 */
export interface CreateResidentInput {
  /**
   * Rent that the new resident needs to pay
   */
  rent: CurrencyInCents;

  /**
   * Incidentals that the new resident needs to pay
   */
  incidentals: CurrencyInCents;

  /**
   * Rent deposit that the resident payed at contract start
   */
  rentDeposit: CurrencyInCents;

  /**
   * First month and year the contract of the new resident starts
   */
  contractStart: MonthYear;

  /**
   * Current water meter count
   */
  waterMeter: number;

  /**
   * Residents included in the contract
   */
  contractResidents: ContractResident[];

  /**
   * Number of residents living in the appartment
   */
  numberOfResidents: number;

  /**
   * Id of the apartment the resident lives in
   */
  apartmentId: string;

  /**
   * Id of the parking space the resident rented
   */
  parkingSpaceId: string | undefined;

  /**
   * Number of apartment keys the resident has
   */
  apartmentKeys: number;

  /**
   * Number of basement keys the resident has
   */
  basementKeys: number;

  /**
   * Number of attic keys the resident has
   */
  atticKeys: number;

  /**
   * Number of front door keys the resident has
   */
  frontDoorKeys: number;

  /**
   * Number of mailbox keys the resident has
   */
  mailboxKeys: number;
}

/**
 * All groups of the form
 */
export type CreateResidentGroups = 'resident' | 'apartment' | 'keys';

/**
 * Config of the edit resident form
 */
export const genericCreateResidentFormConfig: FormConfig<
CreateResidentInput,
CreateResidentGroups
> = {
  formValidationConfig: {
    formValidator: new Validator<CreateResidentInput>({
      rent: ValidationConstraint.Currency,
      incidentals: ValidationConstraint.Currency,
      rentDeposit: ValidationConstraint.Currency,
      contractStart: ValidationConstraint.Defined,
      waterMeter: ValidationConstraint.Defined,
      contractResidents: ValidationConstraint.NoEmptyArray,
      numberOfResidents: ValidationConstraint.Defined,
      apartmentId: ValidationConstraint.Defined,
      apartmentKeys: ValidationConstraint.Defined,
      basementKeys: ValidationConstraint.Defined,
      atticKeys: ValidationConstraint.Defined,
      frontDoorKeys: ValidationConstraint.Defined,
      mailboxKeys: ValidationConstraint.Defined,
    }),
    defaultFormInput: {
      rent: undefined,
      incidentals: undefined,
      rentDeposit: undefined,
      contractStart: new MonthYear(),
      waterMeter: undefined,
      contractResidents: [],
      numberOfResidents: undefined,
      apartmentId: undefined,
      parkingSpaceId: undefined,
      apartmentKeys: undefined,
      basementKeys: undefined,
      atticKeys: undefined,
      frontDoorKeys: undefined,
      mailboxKeys: undefined,
    },
    submitButtonLabel: 'Erstellen',
  },
  formGroupConfig: {
    groupMappings: {
      contractStart: 'resident',
      contractResidents: 'resident',
      numberOfResidents: 'resident',
      rent: 'apartment',
      incidentals: 'apartment',
      rentDeposit: 'apartment',
      waterMeter: 'apartment',
      apartmentId: 'apartment',
      parkingSpaceId: 'apartment',
      apartmentKeys: 'keys',
      basementKeys: 'keys',
      atticKeys: 'keys',
      frontDoorKeys: 'keys',
      mailboxKeys: 'keys',
    },
    groupConfigs: {
      resident: {
        label: 'Mieter',
        icon: {
          component: <PersonIcon />,
          iconPosition: 'start',
        },
      },
      apartment: {
        label: 'Wohnung',
        icon: {
          component: <HomeIcon />,
          iconPosition: 'start',
        },
      },
      keys: {
        label: 'Schl√ºssel',
        icon: {
          component: <KeyIcon />,
          iconPosition: 'start',
        },
      },
    },
    defaultGroup: 'resident',
  },
};

export function getCreateResidentModalConfig(args: {
  emptyApartments: Apartment[];
}): FormConfig<CreateResidentInput, CreateResidentGroups> {
  return {
    ...genericCreateResidentFormConfig,
    formValidationConfig: {
      ...genericCreateResidentFormConfig.formValidationConfig,
      defaultFormInput: {
        ...genericCreateResidentFormConfig.formValidationConfig
          .defaultFormInput,
        apartmentId: args.emptyApartments.at(0)?.id,
      },
    },
  };
}
