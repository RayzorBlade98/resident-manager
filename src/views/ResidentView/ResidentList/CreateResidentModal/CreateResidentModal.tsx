import {
  Button, Grid, InputAdornment, TextField,
} from '@mui/material';
import React, { useState } from 'react';
// eslint-disable-next-line max-len
import GenericModal from '../../../../components/GenericComponents/GenericModal/GenericModal';
// eslint-disable-next-line max-len
import MonthYearInput from '../../../../components/GenericComponents/MonthYearInput/MonthYearInput';
import { ResidentStateManager } from '_/states/saveStates/resident_state';
import { MonthYear, MonthYearUtils } from '_/types/date';
import {
  createResident,
  CreateResidentArguments,
  CreateResidentErrors,
  validateResidentArgs,
} from '_/types/resident';
import { convertCurrencyEurosToCents } from '_/utils/currency';

export interface CreateResidentModalProps {
  /**
   * Defines if the modal is currenly visible or not.
   */
  show: boolean;

  /**
   * Gets called whenever the modal is closed.
   */
  onClose: () => void;
}

/**
 * Modal that contains an input form to create a new resident.
 */
function CreateResidentModal(props: CreateResidentModalProps): JSX.Element {
  // Information of the currently created resident
  const [resident, setResident] = useState<CreateResidentArguments>({
    firstName: '',
    lastName: '',
    rent: null,
    incidentals: null,
    contractStart: MonthYearUtils.getCurrentMonthYear(),
  });
  // Error messages of the input fields
  const [errors, setErrors] = useState<CreateResidentErrors>({});

  /**
   * Creates a function that handles the input of the specified field.
   * It updates the value of the field and applies the validation on it.
   * @param field field of the resident that should be updated
   * @returns function that handles the input of the specified field.
   */
  function residentUpdater(field: string):
  | ((event: React.ChangeEvent<HTMLInputElement>) => void)
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  | ((value: any) => void) {
    /**
     * Updates the value of the specified field of the resident
     * @param value value that should be applied
     */
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    function updateResident(value: any): void {
      const newResident = {
        ...resident,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        [field]: value,
      };
      setResident(newResident);
      setErrors({
        ...errors,
        [field]: validateResidentArgs(
          newResident,
          field as keyof CreateResidentErrors,
        ),
      });
    }

    /**
     * Parses the changed value from the `React.ChangeEvent` and updates the specified field with it.
     * @param event input event of the input component
     */
    function updateResidentWithChangeEvent(
      event: React.ChangeEvent<HTMLInputElement>,
    ): void {
      let value: number | string = event.target.value;
      if (['rent', 'incidentals'].includes(field)) {
        value = convertCurrencyEurosToCents(Number(value));
      }
      updateResident(value);
    }

    // Input for this component doesn't need to be parsed
    if (field === 'contractStart') return updateResident;

    return updateResidentWithChangeEvent;
  }

  /**
   * Handles the saving of the new resident that is triggered whenever the modal is closed.
   */
  function onSave(): void {
    // Validate input
    const newErrors = validateResidentArgs(resident) as CreateResidentErrors;
    const numberOfErrors = Object.keys(newErrors).filter(
      (k) => !!newErrors[k as keyof CreateResidentErrors],
    ).length;

    if (numberOfErrors > 0) {
      // Invalid input(s)
      setErrors(newErrors);
      return;
    }

    // Creates new resident and closes the modal
    ResidentStateManager.addResident(createResident(resident));
    props.onClose();
  }

  return (
    <GenericModal
      title="Neuer Mieter"
      show={props.show}
      onClose={props.onClose}
    >
      {/* Body */}
      <Grid container columnSpacing={2} rowSpacing={2}>
        <Grid item xs={6}>
          <TextField
            id="firstName"
            label="Vorname"
            variant="outlined"
            required
            onChange={residentUpdater('firstName')}
            error={!!errors.firstName}
            helperText={errors.firstName || ''}
          />
        </Grid>
        <Grid item xs={6}>
          <TextField
            id="lastName"
            label="Nachname"
            variant="outlined"
            required
            onChange={residentUpdater('lastName')}
            error={!!errors.lastName}
            helperText={errors.lastName || ''}
          />
        </Grid>
        <Grid item xs={6}>
          <TextField
            id="rent"
            label="Miete"
            type="number"
            required
            onChange={residentUpdater('rent')}
            InputProps={{
              endAdornment: <InputAdornment position="end">€</InputAdornment>,
            }}
            error={!!errors.rent}
            helperText={errors.rent || ''}
          />
        </Grid>
        <Grid item xs={6}>
          <TextField
            id="incidentals"
            label="Nebenkosten"
            type="number"
            required
            onChange={residentUpdater('incidentals')}
            InputProps={{
              endAdornment: <InputAdornment position="end">€</InputAdornment>,
            }}
            error={!!errors.incidentals}
            helperText={errors.incidentals || ''}
          />
        </Grid>
        <Grid item xs={6}>
          <MonthYearInput
            id="contractStart"
            label="Vertragsbeginn"
            priorMonthChoices={6}
            futureMonthChoices={6}
            onChange={
              residentUpdater('contractStart') as (value: MonthYear) => void
            }
          />
        </Grid>
      </Grid>

      {/* Footer */}
      <Button variant="contained" onClick={() => onSave()}>
        Erstellen
      </Button>
    </GenericModal>
  );
}

export default CreateResidentModal;
