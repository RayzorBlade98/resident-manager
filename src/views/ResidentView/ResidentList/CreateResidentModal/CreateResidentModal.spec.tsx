import { act, render } from '@testing-library/react';
import { generateImage } from 'jsdom-screenshot';
import React from 'react';
import { setRecoil } from 'recoil-nexus';
import createResidentState from '../states/create_resident_state';
import CreateResidentModal from './CreateResidentModal';
import MonthYear from '_/extensions/date/month_year.extension';
import ReactTestWrapper from '_/test/ReactTestWrapper';

describe('CreateResidentModal', () => {
  beforeAll(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2023, 4, 31));
  });

  test('should match snapshot', async () => {
    // Arrange
    render(
      <ReactTestWrapper>
        <CreateResidentModal />
      </ReactTestWrapper>,
    );
    act(() => {
      setRecoil(createResidentState, (state) => ({
        ...state,
        showModal: true,
        formValidation: {
          ...state.formValidation,
          formInput: {
            ...state.formValidation.formInput,
            contractStart: new MonthYear(5, 2023),
          },
        },
      }));
    });

    // Assert
    expect(
      await generateImage({ viewport: { width: 650, height: 400 } }),
    ).toMatchImageSnapshot();
  });
});
