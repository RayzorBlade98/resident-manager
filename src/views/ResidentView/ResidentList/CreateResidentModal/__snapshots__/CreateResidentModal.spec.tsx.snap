// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateResidentModal should match snapshot 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r0:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
          id=":r0:"
        >
          Neuer Mieter
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-lc2yqr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </h2>
        <div
          class="MuiDialogContent-root css-uslfu7-MuiDialogContent-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container css-mhc70k-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="firstName"
                  id="firstName-label"
                >
                  Vorname
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id="firstName"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Vorname
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="lastName"
                  id="lastName-label"
                >
                  Nachname
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id="lastName"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Nachname
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="rent"
                  id="rent-label"
                >
                  Miete
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                    id="rent"
                    required=""
                    type="number"
                    value=""
                  />
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                    >
                      €
                    </p>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Miete
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="incidentals"
                  id="incidentals-label"
                >
                  Nebenkosten
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                    id="incidentals"
                    required=""
                    type="number"
                    value=""
                  />
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                    >
                      €
                    </p>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Nebenkosten
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for="contractStart"
                  id="contractStart-label"
                >
                  Vertragsbeginn
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    autocomplete="off"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id="contractStart"
                    inputmode="text"
                    placeholder="MM.JJJJ"
                    required=""
                    type="text"
                    value="06.2023"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-14lo706"
                    >
                      <span>
                        Vertragsbeginn
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-1sszw6i-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Erstellen
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
